text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,596 @@
+1. The lifecycle will compare the list of required mixins to the list of mixins provided by stack and stack buildpacks in accordance with [stage-specific mixin rules](https://github.com/buildpacks/rfcs/pull/109). If any mixins are still not provided, the build will fail.
+1. The lifecycle will run the detect phase for all stackpacks defined in the build-image.
+1. The lifecycle will execute the stack buildpack build phase for all passing stackpack(s). If no stack buildpacks pass detect, the build will continue the build phase as normal (running userspace buildpacks).
+1. After, during, or before the lifecycle's build phase, the lifecycle will begin a new phase called _extend_.
+1. During the extend phase, stack buildpacks that passed detection will run against the run images accordingly (see details below).
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+ A stack buildpack is a special case of buildpack that has the following properties:
+
+* Is run as the `root` user
+* Configured with `privileged = true` in the `buildpack.toml`
+* Can only create one layer
+* May include, in the created layer, modifications to any part of the filesystem, excluding `<layers>`, `<platform>`, and `<cnb>` directories
+* Does not have access to the application source code
+* Is run before all regular buildpacks
+* Is run against both the build and run images
+* Is distributed with the stack run and/or build image
+* May not create layers using the `<layers>` directory
+
+The stackpack interface is similar to the buildpack interface
+* The same `bin/detect` and `bin/build` scripts are required
+* The `bin/detect` will have read-only access to the app
+* The positional arguments for `bin/detect` and `bin/build` are the same (though the values may be different)
+* The environment variables and inputs for `bin/detect` and `bin/build` are the same (though the values may be different)
+
+However, some of the context it is run in is different from regular buildpacks.
+
+For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase excluding the a few specific directories and files.
+
+Additionally, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+A stack can provide stackpacks by including them in the `/cnb/stack/buildpacks` directory, and providing an `/cnb/stack/buildpacks/order.toml` (following the [`order.toml` schema](https://github.com/buildpacks/spec/blob/main/platform.md#ordertoml-toml)) to define their order of execution. The order can be overridden in the `builder.toml` with the following configuration:
```

Comment:
It looks like in the implementation this is `/cnb/stack/order.toml`

Reply:
/cc @jabrown85 what did you decide for this this file?

Reply:
`/cnb/stack/order.toml` is where it is at. I don't mind moving it if we want to. It made sense to me to be one level above the potentially mounted `stack/buildpacks` directory.

Reply:
fixed in 2148aea